### PR TITLE
Revise “Container runtimes” page in light of dockershim removal

### PR DESCRIPTION
**Pull request by PranshuSrivastava**
This updates the container-runtimes [page](https://kubernetes.io/docs/setup/production-environment/container-runtimes/ of the website to include information about the Dockershim deprecation and so that the users can take an informed decision. This is in accordance with #28449

### CODE REVIEW

**Comment by PranshuSrivastava** [comment]
@sftim I had to keep the links this way because they were not being rendered properly if I broke them in between.

**Comment by bradtopol** [comment]
Hi @mikebrow Can you serve as the technical reviewer for this PR? Just add a /lgtm if it is technically correct and I'll take care of the rest.  Thanks. 
/cc @mikebrow 
/assign

**Review by mikebrow — APPROVED** [review]
/lgtm

**Comment by sftim** [comment]
/label tide/merge-method-squash

I'd prefer to omit the whitespace changes, but it's going to be easier I think to merge this and then do fixups.

**Comment by PranshuSrivastava** [comment]
> /label tide/merge-method-squash
> 
> I'd prefer to omit the whitespace changes, but it's going to be easier I think to merge this and then do fixups.

@sftim those whitespace changes are because of my text editor.

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (reylejano) [inline content/en/docs/setup/production-environment/container-runtimes.md:26]: This implies that Docker container runtime is deprecated and that dockershim and Docker container runtime are the same.
I suggest more clarity that dockershim is deprecated

Comment (sftim) [reply content/en/docs/setup/production-environment/container-runtimes.md:26]: @PranshuSrivastava would you be willing to revise the language here?

Comment (PranshuSrivastava) [reply content/en/docs/setup/production-environment/container-runtimes.md]: @sftim @reylejano sorry I missed this review. I have made the change, PTAL.

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (sftim) [reply content/en/docs/setup/production-environment/container-runtimes.md:26]: Rather than “has been”, I recommend “was”. The reason is that “has been” suggests it's a recent removal and we prefer the text of documentation to be more timeless (and easier to maintain).

Comment (PranshuSrivastava) [reply content/en/docs/setup/production-environment/container-runtimes.md]: @mikebrow  @sftim done.

**Review by mikebrow — APPROVED** [review]
LGTM

**Comment by PranshuSrivastava** [comment]
@vincepri @bart0sh PTAL.

**Comment by bart0sh** [comment]
@PranshuSrivastava please, squash your commits. otherwise lgtm.

**Comment by PranshuSrivastava** [comment]
> Hi @PranshuSrivastava, it looks like due to a merge conflict, this PR cannot be merged as of now.
> 
> I would highly suggested you to perform a [rebase](https://kubernetes.io/docs/contribute/new-content/open-a-pr/#merge-conflicts-and-rebasing) (it is a great skill to learn in any source control) on this PR after which you can [squash the commits.](https://kubernetes.io/docs/contribute/new-content/open-a-pr/#merge-conflicts-and-rebasing)
> 
> If you need a hand/help in doing any of this, feel free to ask your questions in the [SIG-docs Slack.](https://kubernetes.slack.com/messages/sig-docs)
> 
> Also if you are not comfortable in doing a _rebase_ of this PR please do tell, in that case any Docs contributor will tell you the way forward .
> 
> Thanks.

@Babapool Done, thanks.

**Comment by PranshuSrivastava** [comment]
> @PranshuSrivastava please, squash your commits. otherwise lgtm.

@bart0sh done

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (tengqm) [inline content/en/docs/setup/production-environment/container-runtimes.md:34]:
```suggestion
/blog/2020/12/02/dockershim-faq/)
```

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (tengqm) [inline content/en/docs/setup/production-environment/container-runtimes.md:36]: We use relative paths rather than absolute ones. Also please help make sure the link appears on one line.
```suggestion
/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)
```

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (tengqm) [inline content/en/docs/setup/production-environment/container-runtimes.md:40]:
```suggestion
/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/)
```

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (tengqm) [inline content/en/docs/setup/production-environment/container-runtimes.md:43]:
```suggestion
dockershim you can follow [the migration guide](/docs/tasks/administer-cluster/migrating-from-dockershim/)
```

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (tengqm) [inline content/en/docs/setup/production-environment/container-runtimes.md:44]:
```suggestion
to migrate from dockershim.
```

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (tengqm) [inline content/en/docs/setup/production-environment/container-runtimes.md:434]:
```suggestion
On each of your nodes, install the Docker for your Linux distribution as per
```

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (tengqm) [inline content/en/docs/setup/production-environment/container-runtimes.md:449]:
```suggestion
{{< note >}}
`overlay2` is the preferred storage driver for systems running Linux kernel version 4.0 or higher,
or RHEL or CentOS using version 3.10.0-514 and above.
{{< /note >}}
```

**Comment by PranshuSrivastava** [comment]
Done @tengqm

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (reylejano) [inline content/en/docs/setup/production-environment/container-runtimes.md:33]:
```suggestion
Docker container runtime (Dockershim) has been [deprecated](
```
Implies that dockershim and Docker container runtime are the same, please revise

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (reylejano) [inline content/en/docs/setup/production-environment/container-runtimes.md:38]:
```suggestion
and is planned to be removed in by 1.24, therefore to use Docker you need to
be running an older version of Kubernetes.
```
This is not true, you can use cri-dockerd and use Docker

**Comment by sftim** [comment]
Thanks for working on this @PranshuSrivastava 

It's an important topic so we're keen to get the nuances and technical details exactly right.

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (reylejano) [reply content/en/docs/setup/production-environment/container-runtimes.md]: @PranshuSrivastava check the links from the preview, @tengqm 's suggestion is correct

Comment (PranshuSrivastava) [reply content/en/docs/setup/production-environment/container-runtimes.md]: @reylejano Okay

**Comment by PranshuSrivastava** [comment]
@reylejano @sftim PTAL

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (reylejano) [inline content/en/docs/setup/production-environment/container-runtimes.md:432]:
```suggestion
On each of your nodes, install Docker for your Linux distribution as per
```

Comment (PranshuSrivastava) [reply content/en/docs/setup/production-environment/container-runtimes.md:432]: done @reylejano

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (PranshuSrivastava) [reply content/en/docs/setup/production-environment/container-runtimes.md]: @reylejano PTAL

Comment (mikebrow) [reply content/en/docs/setup/production-environment/container-runtimes.md]: you can use the containerd release tar provided at containerd.io/downloads.. but that link does not come with an installer or install instructions... more would be needed or linked to in the docs to make that make sense/work (what to do after extracting from the tar for example)

Comment (afbjorklund) [reply content/en/docs/setup/production-environment/container-runtimes.md]: It's the same story as cri-dockerd, except that doesn't even have downloads...
(I will pretend that the cri-o instructions are easy to follow and set an example.)

* https://github.com/kubernetes/kubernetes/issues/106917#issuecomment-1023964814

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (sftim) [inline content/en/docs/setup/production-environment/container-runtimes.md:42]: If this text is present, the PR must target the _dev-1.24_ branch and not _main_.

/hold
until that discrepancy is dealt with.

Comment (PranshuSrivastava) [reply content/en/docs/setup/production-environment/container-runtimes.md:42]: @sftim done

Comment (sftim) [reply content/en/docs/setup/production-environment/container-runtimes.md:42]: Thanks
/unhold

however: we may well want a similar-but-different change targeting _main_, so that we can blog about this before the release of v1.24

This PR can still go ahead, I think.

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (afbjorklund) [reply content/en/docs/setup/production-environment/container-runtimes.md]: I used the normal k8s packages, that installed crictl 1.19 for me

If you want the latest/greatest, you can go for the ones at Kubic

Comment (afbjorklund) [reply content/en/docs/setup/production-environment/container-runtimes.md]: @reylejano here are the package dependencies:

Depends: kubelet (>= 1.19.0), kubectl (>= 1.19.0), kubernetes-cni (>= 0.8.7), **cri-tools (>= 1.19.0)**

kubeadm/kubernetes-xenial 1.23.3-00 amd64
kubelet/kubernetes-xenial 1.23.3-00 amd64
kubectl/kubernetes-xenial 1.23.3-00 amd64
kubernetes-cni/kubernetes-xenial 0.8.7-00 amd64
cri-tools/kubernetes-xenial 1.19.0-00 amd64

https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#installing-kubeadm-kubelet-and-kubectl

You still need to configure /etc/crictl.yaml manually

And to make sure to add a CNI plugin (I used flannel)

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (afbjorklund) [inline content/en/docs/setup/production-environment/container-runtimes.md:435]: I don't think Kubernetes validates against Docker anymore, and there doesn't seem to be a version in that file.

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (afbjorklund) [inline content/en/docs/setup/production-environment/container-runtimes.md:167]: There are no "containerd.io" packages on the containerd.io website, there it just called "containerd" (and no packages)

The alternative (to "containerd") would be "cri-containerd-cni" which also includes the CRI and CNI configuration:

`/etc/crictl.yaml`

`etc/cni/net.d/10-containerd-net.conflist`

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (afbjorklund) [inline content/en/docs/setup/production-environment/container-runtimes.md:168]: You don't need to set up the **Docker** package repository, if installing from the containerd.io.

So either you install the "containerd.io" package from Docker, or you install the "containerd"
package from your distribution (`apt install containerd`) or download it directly from containerd.io

If you want to keep both options (docker.com and containerd.io), they need to be clearly separated.

**Comment by reylejano** [comment]
/remove-language es id it pt ru uk zh

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (bart0sh) [inline content/en/docs/setup/production-environment/container-runtimes.md:39]: Leading "to migrate from dockershim" looks redundant/tautological to me:

For migrating from dockershim ... **to migrate from dockershim.**

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (bart0sh) [reply content/en/docs/setup/production-environment/container-runtimes.md:168]: +1

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (PranshuSrivastava) [reply content/en/docs/setup/production-environment/container-runtimes.md]: done @bart0sh

**Comment by PranshuSrivastava** [comment]
@vincepri PTAL

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (sftim) [reply content/en/docs/setup/production-environment/container-runtimes.md:167]: Where does the package name `cri-containerd-cni` come from?

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (sftim) [inline content/en/docs/setup/production-environment/container-runtimes.md:42]:
```suggestion
Kubernetes releases before v1.24 included a direct integration with Docker Engine,
using a component named _dockershim_. That special direct integration is no longer
part of Kubernetes (this removal was
[announced](/blog/2020/12/08/kubernetes-1-20-release-announcement/#dockershim-deprecation)
as part of the v1.20 release).

You can read
[Check whether Dockershim deprecation affects you](/docs/tasks/administer-cluster/migrating-from-dockershim/check-if-dockershim-deprecation-affects-you/) to understand how this removal might
affect you. To learn about migrating from using dockershim, see
[Migrating from dockershim](/docs/tasks/administer-cluster/migrating-from-dockershim/).
```

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (bart0sh) [reply content/en/docs/setup/production-environment/container-runtimes.md:167]: @sftim cri-containerd-cni is a tarball, recommended as a way to install by the containerd documentation: https://containerd.io/docs/getting-started/#starting-containerd

+1 for not mentioning containerd.io package.

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (bart0sh) [reply content/en/docs/setup/production-environment/container-runtimes.md:168]: @afbjorklund thank you.

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (sftim) [inline content/en/docs/setup/production-environment/container-runtimes.md:448]:
```suggestion
You can also find third-party adapters that let you use Docker Engine with Kubernetes
through the supported {{< glossary_tooltip term_id="cri" text="Container Runtime Interface">}}
(CRI).

The following CRI adapters are designed to integrate Docker Engine with CRI clients (such
as Kubernetes v{{< skew currentVersion >}}):

- [`cri-dockerd`](https://github.com/Mirantis/cri-dockerd) from Mirantis

{{< note >}}
If you use Docker Engine on a node that is running Linux kernel version 4.0 or higher, or that
runs RHEL or CentOS and uses the Linux kernel version 3.10.0-514 and above,
the Kubernetes project suggests using the `overlay2` storage driver.
{{< /note >}}
```

**Review by sftim — CHANGES_REQUESTED** [review]
It looks like this needs some more changes - I've added inline feedback.

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (neolit123) [reply content/en/docs/setup/production-environment/container-runtimes.md:167]: makes sense to link to https://containerd.io/downloads/ and https://containerd.io/docs/getting-started/#starting-containerd

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (neolit123) [reply content/en/docs/setup/production-environment/container-runtimes.md:168]: we should remove the docker repository from the containerd steps and leave the user two options:
- following the guide in https://containerd.io/downloads/ and https://containerd.io/docs/getting-started/#starting-containerd
- use the package manager of the Linux distro

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (bart0sh) [reply content/en/docs/setup/production-environment/container-runtimes.md]: @PranshuSrivastava thanks, looks better now.

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (bart0sh) [reply content/en/docs/setup/production-environment/container-runtimes.md:167]: @afbjorklund installing CNI can cause [hard to debug issues](https://github.com/cri-o/cri-o/issues/2885). I'd recommend to install containerd from `cri-containerd` tarball instead.

Comment (afbjorklund) [reply content/en/docs/setup/production-environment/container-runtimes.md:167]: Kubernetes only works with one CNI at a time, installing podman or cri-o or even containerd is known to interfere with cluster setup...

The recommendation is to wipe /etc/cni before configuring and starting the cluster

Comment (bart0sh) [reply content/en/docs/setup/production-environment/container-runtimes.md:167]: @afbjorklund  So, what's the point to install CNI if it's going to be removed anyway?

btw, I don't see any recommendation to remove CNI in [this documentation](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/). Can you point me out to the recommendation you've mentioned?

Comment (afbjorklund) [reply content/en/docs/setup/production-environment/container-runtimes.md:167]: I don't think there is any instructions regarding CNI at all ? So it was mostly how I was adding myself.

`rm -f /etc/cni/net.d/*.conf*`

`kubectl apply -f https://raw.githubusercontent.com/flannel-io/flannel/v0.14.0/Documentation/kube-flannel.yml`

This of course also _breaks_ all other usage, so you need to move the podman configuration directory.

```
# Path to the directory where CNI configuration files are located.
#
# network_config_dir = "/etc/cni/net.d/"
```

Similar with the nerdctl configuration, so that it doesn't interfer with Kubernetes' containerd config (of CNI) ?

Comment (afbjorklund) [reply content/en/docs/setup/production-environment/container-runtimes.md:167]: @bart0sh I am not sure what the best way is to package this.

The user _needs_ to set up /etc/crictl.yaml and /etc/cni/net.d

But at the same time, there can be only one container runtime and only one container networking for Kubernetes.

So the files from one will conflict with the files from the other ones. Maybe they should go in /usr/share or in docs ?

Comment (bart0sh) [reply content/en/docs/setup/production-environment/container-runtimes.md:167]: @afbjorklund installing https://github.com/containerd/containerd/releases/download/v1.6.0/cri-containerd-1.6.0-linux-arm64.tar.gz and then following [this guide](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/) is the most straightforward way to get working kubernetes cluster in my opinion.

Comment (afbjorklund) [reply content/en/docs/setup/production-environment/container-runtimes.md:167]: Sure, but it doesn't have any details ? It says you need a CRI and you need a CNI, but after that it is just "GL;HF".

The entire process was more like: https://github.com/lima-vm/lima/blob/v0.8.3/examples/k8s.yaml#L28_L79

Comment (bart0sh) [reply content/en/docs/setup/production-environment/container-runtimes.md:167]: [this guide](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/) was referencing [container runtimes](https://kubernetes.io/docs/setup/production-environment/container-runtimes/) previously IIRC. So, I always followed first guide and it worked for me.

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (afbjorklund) [reply content/en/docs/setup/production-environment/container-runtimes.md:168]: Would it be possible to supply deb packages for containerd, the same way as for "cri-tools" and "kubernetes-cni" ?

From apt.kubernetes.io, that is.

Comment (neolit123) [reply content/en/docs/setup/production-environment/container-runtimes.md:168]: The package management in k8s needs to be community owned. Currently Google still holds the keys. But also some reorganization of the package repositories might be required for better version flexibility.

The containerd maintainers can open a discussion about that with the k8s SIG Release. In the meantime we should document the best option, IMO.

Comment (afbjorklund) [reply content/en/docs/setup/production-environment/container-runtimes.md:168]: Maybe CRI and CNI should be decoupled from the packaging as well, they would be less outdated that way I think.

Comment (sftim) [reply content/en/docs/setup/production-environment/container-runtimes.md:168]: @PranshuSrivastava I recommend following the approach outlined in https://github.com/kubernetes/website/pull/30882#discussion_r816261724

@afbjorklund, I'm keen to keep the focus of this PR on just the set of changes we'd make to this page so that the v1.24 docs make sense after the dockershim component is removed. Once this merges, we can consider follow-up PRs that make further improvements.

As the proverb doesn't quite say, the perfect is the enemy of the merged.

Comment (afbjorklund) [reply content/en/docs/setup/production-environment/container-runtimes.md:168]: I think changing from docker.com to containerd.io would be enough, go with the tarball if that is all there is.

Comment (bart0sh) [reply content/en/docs/setup/production-environment/container-runtimes.md:168]: Can we keep containerd installation instructions as they are as they're not actually related to the purpose of this PR. We can update them in a separate PR I believe.

Comment (sftim) [reply content/en/docs/setup/production-environment/container-runtimes.md:168]: After we remove the dockershim, the current instructions are too likely to confuse people. I think fixing that does align with the intent of this PR.

**Comment by sftim** [comment]
/retitle Revise “Container runtimes” page in light of dockershim removal

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (mikebrow) [reply content/en/docs/setup/production-environment/container-runtimes.md:42]: SGTM

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (mikebrow) [reply content/en/docs/setup/production-environment/container-runtimes.md:448]: SGTM

File: content/en/docs/setup/production-environment/container-runtimes.md
Comment (afbjorklund) [reply content/en/docs/setup/production-environment/container-runtimes.md:167]: Minor packaging detail: runc is bundled with "containerd.io" and with "cri-containerd", there is no "runc.io" package.

There is a regular system "runc" package, that is a dependency of the regular system "containerd" package though.

The docker packages are basically internal to docker, they could change (and have changed before) at any time.

Haven't looked at the rpm packaging (this was deb), but assume there is a similar vendor vs. system duplication ?

Comment (sftim) [reply content/en/docs/setup/production-environment/container-runtimes.md:167]: For SIG Docs: we want to be as generic as we can be about third-party solutions. Even if someone uses containerd, we don't know what OCI runtime they prefer. `runc` is common but there are others that are no less valid.

Feedback on this PR should be aimed at helping the author, @PranshuSrivastava, get this ready for merge. For any other discussion, use issues or Slack.

Also:
```
$ dpkg-query --show runc
runc	1.0.1-0ubuntu2~20.04.1
```
(that's a `.deb`)


@PranshuSrivastava the feedback I'd highlight for you here is: https://github.com/kubernetes/website/pull/30882#discussion_r816258989

**Comment by bradtopol** [comment]
Assigning @sftim as the reviewer since he is the one doing the in depth review here.
/unassign
/assign @sftim

**Comment by sftim** [comment]
@PranshuSrivastava would you have time to revise this further?

Another option that we have is to take your existing work and build upon it, merging a different PR that is based on what's here.

**Comment by reylejano** [comment]
@PranshuSrivastava , are you able to continue to review the comments and make revisions to progress this PR?

If not, I agree with @sftim 's [comment](https://github.com/kubernetes/website/pull/30882#issuecomment-1077867090)

**Comment by PranshuSrivastava** [comment]
@reylejano @sftim Yes I have time to work further on this PR. I have addressed all the comments so far and will continue to work on the reviews provided in the following days.

**Pull request merged**
